p4-16/psa/PSA.mdk

```diff
@@ -2249,6 +2249,64 @@ The control plane should provide a
 
 ```
 
+
+# Atomicity of control plane API operations
+
+All table add, delete, and modify operations must be atomic relative
```

Comment:
Since this is a recommendation, perhaps we should replace the word must with "should". There are so many situations I've encountered in the real life, where this is "kinda" the rule, but there are exceptions. And in many cases they are not fatal -- all the control plane needs to do is to make sure that this entry that is being changed in some non-atomic fashion doesn't get hit during that process. There are many ways to do it.

Reply:
Are these only recommendations?  I thought the intent is that these were going to be requirements for PSA implementations.

Reply:
Indeed, my recollection is also that we are going to require atomic semantics for the table operations. I do not think we should water down this requirement.

Reply:
I'm not sure if this is worth mentioning in the PSA spec, but realize that a PSA implementation _could_ be done where "simple enough" tables were automatically atomic w.r.t. single control plane operations as proposed here, but "compex cases" (whatever that means for an implementation, e.g. too many bits of action parameters, or keys have too many bits, etc.) would by default be rejected as non-atomic by the compiler, but the implementation could also provide table annotations where the P4 programmer could explicitly say "I am OK if this table doesn't have atomic control plane operations, because my control plane code is written to handle that."

Because a control plane might need to be written in a significantly different way for a table that didn't provide this atomicity guarantee, vs. one that did, it seems much safer to require the programmer to "opt in" for the less safe behavior, than for the default to go the other way.